icu4c/source/common/unicode/unistr.h

```diff
@@ -3485,7 +3658,17 @@ class U_COMMON_API UnicodeString : public Replaceable
    * Internal string contents comparison, called by operator==.
    * Requires: this & text not bogus and have same lengths.
    */
-  UBool doEquals(const UnicodeString &text, int32_t len) const;
+  inline UBool doEquals(const UnicodeString &text, int32_t len) const {
+    return doEquals(text.getArrayStart(), len);
+  }
+  UBool doEquals(const char16_t *text, int32_t len) const;
+#ifdef MAYBE_EQUALS_PTR
+  // TODO: This could be a useful optimization, by avoiding u_strlen(text).
+  // However, it is possible that constructing a std::u16string_view is even better
+  // when the text is a string literal, since the compiler should be able to
+  // determine the length at compile time.
+  UBool doEquals(const char16_t *text) const;
+#endif  // MAYBE_EQUALS_PTR
```

Comment:
I had originally added this overload for disambiguation.

Then I realized that constructing a UnicodeString or a string_view from a NUL-terminated string calls strlen() which is wasteful when comparing for equality. So I optimized it.

However, I believe that when a string_view is constructed from a string literal, the length is determined at compile time, so creating a string_view and letting the other new overload take it would be fastest in a common use case.

So I plan to remove this again because (a) we no longer need it for the stated purpose of this PR and (b) as an optimization it may backfire.

WDYT?

Reply:
> I believe that when a string_view is constructed from a string literal, the length is determined at compile time

I am not sure about that, but I would remove it anyway; if we have the templatized overload, we can be as smart as we want later on within that API surface.

(I think to determine the length at compile time you would need an overload templatized on the length that takes a reference to a char16_t array. That is fun, we can absolutely do it, but it arguably would not be an API change anymore, and it seems to be a job for another PR.)

Reply:
I am not proposing (at this time) that we create our own overloads that take an array.
I thought that the compiler was doing this already, but I can't find a reference to that.
I removed this from-pointer overload.

Also, there is another way:
C++17 also added a string literal syntax (constexpr) for std::u16string_view, and that definitely determines the length at compile time. (Requires a "using" directive.)
https://en.cppreference.com/w/cpp/string/basic_string_view/operator%22%22sv
I tried it out and added a little documentation.
We should change the ICU library code to use this style.

Reply:
Actually, considering the declaration
```
constexpr basic_string_view( const CharT* s );
```
How could this be a constexpr and not determine the string length at compile time?

Reply:
Right, it can do the whole thing at compile time (but then the compiler is just doing a strlen).
Note that you can distinguish a thing that takes the length of the literal from one that does strlen, by passing `"something\0like this."`.